lib/request.ts

```diff
@@ -47,23 +47,41 @@ export interface TimingPhases {
   total?: number;
 }
 
-export class ErrorWithTimings extends Error {
-  constructor(
-    originalError: Error,
-    public timings: Timings,
-    public timingPhases: TimingPhases
-  ) {
-    super(originalError.message);
-  }
-}
-
 const subtract = (a?: number, b?: number): number | undefined => {
   if (typeof a === "number" && typeof b === "number") {
     return a - b;
   }
   return undefined;
 };
 
+export class RequestError extends Error {
+  public timings?: Timings;
+  public timingPhases?: TimingPhases;
+  public requestOptions: ServiceClientRequestOptions;
+  constructor(
+    message: string,
+    requestOptions: ServiceClientRequestOptions,
+    timings?: Timings
+  ) {
+    super(message);
+    this.requestOptions = requestOptions;
+    this.timings = timings;
+    this.timingPhases = timings && makeTimingPhases(timings);
+  }
+}
+
+export class ConnectionTimeoutError extends RequestError {
+  constructor(requestOptions: ServiceClientRequestOptions, timings?: Timings) {
+    super("socket timeout", requestOptions, timings);
+  }
+}
+
+export class UserTimeoutError extends RequestError {
```

Comment:
Discussed in-person: don't extend here, this might cause issues when people will start doing `if instanceof`s